### PR TITLE
Camera rotation subsystem

### PR DESCRIPTION
**Pull request by BeardyKing**
debug editor camera controller
+ delta time fix

### CODE REVIEW

File: core/src/forward_renderer.cpp [hunk 1]
```diff
@@ -166,9 +246,167 @@ void ForwardRenderer::color_pass() {
...
+
+        // Set view and projection matrix for view 0.
+        {
+            view = glm::lookAt(pos, lookTarget, up);
+            glm::mat4 proj = glm::perspective(fovY, aspectRatio, zNear, zFar);
+            bgfx::setViewTransform(idx, &view[0][0], &proj[0][0]);
+        }
+    }
+
+    auto entities = registry.view<Transform, InstanceMesh, Material, Name, PostProcessing>();
```
Comment (BertTDev) [inline core/src/forward_renderer.cpp:292]: So all GUI entities will need postprocessing components?

**Review by BertTDev — CHANGES_REQUESTED** [review]

File: core/src/camera_rotation.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        m_yaw = rotation.z;
+    }
+}
+
+void CameraRotation::on_update(double m_delta_time) {
+    //= CHECK VALID INPUT ==
+    InputManager m_inputManager = m_engine.get_window_module().lock()->get_input_manager();
+
+    vec2d currentMousePos = m_inputManager.get_absolute_position();
+    if (glm::isinf(currentMousePos.x) || glm::isinf(currentMousePos.y)) {
```
Comment (lmariscal) [inline core/src/camera_rotation.cpp:48]: We don't need to use the `glm::` namespace specifier.

File: core/src/camera_rotation.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        return;
+    }
+    m_mouseDelta = (currentMousePos - m_lastMousePosition);
+
+    //= 'E' TOGGLE MOUSE HIDDEN ==
+    camera_key_input();
+
+    using namespace components;
```
Comment (lmariscal) [inline core/src/camera_rotation.cpp:56]: If you could please put this just after the function signature.

File: core/src/camera_rotation.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+        m_roll = rotation.x;
+        m_pitch = rotation.y;
+        m_yaw = rotation.z;
+    }
+}
+
+void CameraRotation::on_update(double m_delta_time) {
+    //= CHECK VALID INPUT ==
+    InputManager m_inputManager = m_engine.get_window_module().lock()->get_input_manager();
```
[the remark this answers is not in the document]
Comment (BeardyKing) [reply core/src/camera_rotation.cpp:45]: I agree but currently the return type is a const reference and when caching it it breaks the ubuntu build and casting it to a shared_ptr breaks it too.
the whole input system should probably be separated from window and become it's own sub_system so we can have a simple shared_ptr as a reference

File: core/src/forward_renderer.cpp (hunk 1 above)
Comment (BeardyKing) [reply core/src/forward_renderer.cpp:292]: this is some redundant example code, anything that happens inside of the "gui_pass" using the "idx" provided by the function will be drawn to the GUI Framebuffer.

the framebuffer is then provided alongside the (colour / geometry) pass as a texture for the post processing stack where anything in the GUI pass that has an alpha > 0 will be rendered over the top of the (colour / geometry) texture

File: core/src/camera_rotation.cpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        look.y = sinf(radians(eulerRotation.y));
+        look.z = cosf(radians(eulerRotation.y)) * cosf(radians(eulerRotation.z));
+        m_forward = glm::normalize(look);
+
+        //= CALCULATE LOCAL RIGHT AND UP == // TODO STORE / CALC THESE IN TRANSFORM
+        m_right = glm::normalize(glm::cross(m_forward, vec3(0, 1, 0)));
+        m_up = glm::normalize(glm::cross(m_right, m_forward));
+
+        //= CAMERA MOVEMENT
+        vec3 vecDeltaTime = vec3(m_delta_time);
```
[the remark this answers is not in the document]
Comment (BeardyKing) [reply core/src/camera_rotation.cpp:102]: valid

File: core/src/camera_rotation.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        GameObject go = goOpt.value();
+        Transform& transform = go.get_component<Transform>();
+
+        auto rotation = transform.get_rotation_euler();
+
+        m_roll = rotation.x;
+        m_pitch = rotation.y;
+        m_yaw = rotation.z;
```
Comment (lmariscal) [inline core/src/camera_rotation.cpp:41]:
```suggestion
        m_pitch = rotation.x;
        m_yaw = rotation.y;
        m_roll = rotation.z;
```

File: core/src/forward_renderer.cpp (hunk 1 above)
Comment (BertTDev) [reply core/src/forward_renderer.cpp:292]: This makes sense thank you!

File: core/src/camera_rotation.cpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+        transform.set_rotation_euler(vec3(m_roll, m_pitch, m_yaw));
+        auto eulerRotation = transform.get_rotation_euler();
+
+        //= CALCULATE FORWARD VECTOR == // TODO STORE / CALC THESE IN TRANSFORM
+        glm::vec3 look;
+        look.x = cosf(radians(eulerRotation.y)) * sinf(radians(eulerRotation.z));
+        look.y = sinf(radians(eulerRotation.y));
+        look.z = cosf(radians(eulerRotation.y)) * cosf(radians(eulerRotation.z));
+        m_forward = glm::normalize(look);
```
Comment (lmariscal) [inline core/src/camera_rotation.cpp:95]:
```suggestion
        //= CAMERA ROTATION
        m_roll = 0.0f;
        m_pitch += ((float)m_mouseDelta.y * (float)m_mouseSensitivity.y) * (float)m_delta_time;
        m_yaw += ((float)-m_mouseDelta.x * (float)m_mouseSensitivity.x) * (float)m_delta_time;

        m_pitch = clamp(m_pitch, -89.f, 89.f);
        transform.set_rotation_euler(vec3(m_pitch, m_yaw, m_roll));

        //= CALCULATE FORWARD VECTOR == // TODO STORE / CALC THESE IN TRANSFORM
        vec3 look;
        look.x = sinf(radians(m_yaw));
        look.y = cosf(radians(m_pitch));
        look.z = sinf(radians(m_pitch)) + cosf(radians(m_yaw));

        m_forward = normalize(look);
```

File: core/src/camera_rotation.cpp [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    }
+    if (m_inputManager.key_on_release(KeyCode::D)) {
+        movementDirectionXYZ += m_right;
+    }
+    if (m_inputManager.key_on_release(KeyCode::R)) {
+        movementDirectionXYZ += m_up;
+    }
+    if (m_inputManager.key_on_release(KeyCode::F)) {
+        movementDirectionXYZ += -m_up;
+    }
```
Comment (lmariscal) [inline core/src/camera_rotation.cpp:148]: For some weird reason the proper callbacks to update the input is being removed from some branches, this should be inside `window.cpp` in `on_update`.

These should be `key_pressed` not `key_on_release`

**Review by lmariscal — CHANGES_REQUESTED** [review]

File: core/src/camera_rotation.cpp (hunk 2 above)
Comment (BeardyKing) [reply core/src/camera_rotation.cpp:48]: isinf is a function in corectr_math.h and glm

**Review by BertTDev — APPROVED** [review]

**Review by BertTDev — APPROVED** [review]

**Review by BeardyKing — COMMENTED** [review]
lgtm

**Review by lmariscal — APPROVED** [review]

**Pull request merged**